### PR TITLE
created proper authentication as per issue #19

### PR DESCRIPTION
**Pull request by gregtuc**
Used bcrypt to hash passwords and JSON Web Tokens that the client will be able to used for login/registration.

Endpoints were changed to having a prefix of /api/.

Database connection now uses mongoose, which enables a variety of benefits including using the mongoose Schema. This is more of a change so that the project follows proper coding convention and popular technique. 

Also added an endpoint of /api/currentuser. When the client sends a GET request to this endpoint containing their JWT, a response with user details will be sent. Middleware was created for this which authenticates the JWT and returns a status 500 if the request is invalid.

### CODE REVIEW

File: server/database/default.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-  "ConnectionUrl": "mongodb+srv://<user>:<pass>@cluster0-pf0b2.mongodb.net/test?retryWrites=true&w=majority"
+  "ConnectionUrl": "mongodb+srv://greg34910:cookie123@cluster0-pf0b2.mongodb.net/memestagram?retryWrites=true&w=majority"
```
Comment (ginywiny) [inline server/database/default.json]: Would it be possible to make this line a generic for the database?

**Comment by ginywiny** [comment]
I suggest you make a separate pull request for the travis-ci build, since there could be many additional files and changes to get it to work.

**Comment by gregtuc** [comment]
> I suggest you make a separate pull request for the travis-ci build, since there could be many additional files and changes to get it to work.

Makes sense, I'm gonna make like a separate travis branch on my forked repo to keep travis from spamming it too much moving forward.

**Comment by ginywiny** [comment]
> > I suggest you make a separate pull request for the travis-ci build, since there could be many additional files and changes to get it to work.
> 
> Makes sense, I'm gonna make like a separate travis branch on my forked repo to keep travis from spamming it too much moving forward.

Keep me updated with this.

**Comment by cottonswab** [comment]
Greg can you please improve your comments on this?

**Pull request merged**